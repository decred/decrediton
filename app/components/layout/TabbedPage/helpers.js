import { createElement as h, cloneElement as k } from "react";
import { matchPath } from "react-router-dom";
import { isArray } from "lodash";
import { spring } from "react-motion";
import theme from "theme";
import TabbedPageTab from "./TabbedPageTab";

const getTabs = (children) => {
  if (!isArray(children)) children = [children];
  return children
    .filter((c) => c.type === TabbedPageTab)
    .map((c, i) => ({ index: i, tab: c }));
};

export const getMatchedTab = (location, children) => {
  const tabs = getTabs(children);
  return tabs.find(
    (t) => !!matchPath(location.pathname, { path: t.tab.props.path })
  );
};

export const getStyles = (matchedTab) => {
  if (!matchedTab) {
    return [];
  }

  const element = React.isValidElement(matchedTab.tab.props.component)
    ? k(matchedTab.tab.props.component, {
        ...matchedTab.tab.props,
        ...matchedTab.tab.props.component.props
      })
    : // If the component props are needed, make a valid react element
      // before send, otherwise they will be undfined.
      h(matchedTab.tab.props.component, { ...matchedTab.tab.props }, null);
  return [
    {
      key: matchedTab.tab.props.path,
      data: { matchedTab, element },
      style: { left: spring(0, theme("springs.tab")), opacity: 1 }
    }
  ];
};

export const willLeave = (dir) => {
  const pos = dir === "l2r" ? -1000 : +1000;
  return {
    left: spring(pos, { stiffness: 180, damping: 20 }),
    opacity: spring(0)
  };
};

export const willEnter = (dir) => {
  const pos = dir === "l2r" ? +1000 : -1000;
  return { left: pos, opacity: 1 };
};
