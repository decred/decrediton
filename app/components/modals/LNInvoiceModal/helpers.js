import { FormattedMessage as T } from "react-intl";
export const getInvoiceDetails = (invoice, tsDate) => {
  const details = [
    {
      label: <T id="ln.invoicesModal.hash" m="Hash" />,
      value: invoice?.rHashHex,
      copyable: true,
      truncate: 40
    },
    {
      label: <T id="ln.invoicesModal.desc" m="Description" />,
      value: invoice?.memo
    }
  ];

  invoice?.htlcsList?.forEach((htlc, index) => {
    const response = {
      label: (
        <>
          <T id="ln.invoicesModal.htlc" m="HTLC" /> {index}
        </>
      ),
      value: [
        {
          label: <T id="ln.invoicesModal.htlc.state" m="State" />,
          value: htlc.state
        },
        {
          label: <T id="ln.invoicesModal.htlc.chanId" m="Channel id" />,
          value: htlc.chanId
        },
        {
          label: (
            <T id="ln.invoicesModal.htlc.acceptHeight" m="Accept Height" />
          ),

          value: htlc.acceptHeight
        },
        {
          label: (
            <T id="ln.invoicesModal.htlc.acceptTimeLabel" m="Accept Time" />
          ),
          value: (
            <T
              id="ln.invoicesModal.htlc.acceptTime"
              m="{acceptTime, date, medium} {acceptTime, time, short}"
              values={{ acceptTime: tsDate(htlc.acceptTime) }}
            />
          )
        },
        {
          label: (
            <T id="ln.invoicesModal.htlc.expiryHeight" m="Expiry Height" />
          ),

          value: htlc.expiryHeight
        },
        {
          label: (
            <T id="ln.invoicesModal.htlc.resolveTimeLabel" m="Resolve Time" />
          ),
          value: (
            <T
              id="ln.invoicesModal.htlc.resolveTime"
              m="{resolveTime, date, medium} {resolveTime, time, short}"
              values={{ resolveTime: tsDate(htlc.resolveTime) }}
            />
          )
        }
      ]
    };
    details.push(response);
  });

  return details;
};
