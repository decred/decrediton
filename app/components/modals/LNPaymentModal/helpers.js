import { FormattedMessage as T } from "react-intl";
import { Balance } from "shared";
export const getPaymentDetails = (payment) => {
  const details = [
    {
      label: <T id="ln.paymentModal.hash" m="Hash" />,
      value: payment?.paymentHash,
      copyable: true,
      truncate: 40
    }
  ];

  if (payment?.description) {
    details.push({
      label: <T id="ln.paymentModal.desc" m="Description" />,
      value: payment.description
    });
  }

  if (payment?.destination) {
    details.push({
      label: <T id="ln.paymentModal.destination" m="Destination" />,
      value: payment.destination,
      truncate: 40
    });
  }

  payment?.htlcsList?.forEach((htlc, index) => {
    const response = {
      label: (
        <>
          <T id="ln.paymentModal.htlc" m="HTLC" /> {index}
        </>
      ),
      value: [
        {
          label: <T id="ln.paymentModal.htlc.status" m="Status" />,
          value: htlc.status
        },
        {
          label: <T id="ln.paymentModal.htlc.totalAmt" m="Total Amount" />,
          value: <Balance amount={htlc.route.totalAmt} />
        },
        {
          label: <T id="ln.paymentModal.htlc.totalFees" m="Total Fees" />,
          value: <Balance amount={htlc.route.totalFees} />
        }
      ]
    };
    // hopList
    htlc.route?.hopsList?.forEach((hop, hopIndex) => {
      const hopResponse = {
        label: (
          <>
            <T id="ln.paymentModal.routeHop" m="Hop" /> {hopIndex}
          </>
        ),
        value: [
          {
            label: <T id="ln.paymentModal.htlc.hop.fee" m="Fee" />,
            value: <Balance amount={hop.fee} />
          },
          {
            label: <T id="ln.paymentModal.htlc.hop.pubkey" m="PubKey" />,
            value: hop.pubKey,
            copyable: true,
            truncate: 20
          }
        ]
      };
      response.value.push(hopResponse);
    });

    details.push(response);
  });

  return details;
};
