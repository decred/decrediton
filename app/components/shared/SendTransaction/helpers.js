export const baseOutput = () => ({
  key: "output_0",
  data: {
    destination: "",
    amount: null,
    error: { address: null, amount: null }
  }
});
