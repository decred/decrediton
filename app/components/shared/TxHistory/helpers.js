import { FormattedMessage as T } from "react-intl";
import * as txTypes from "constants/Decrediton";

export const messageByType = {
  [txTypes.TICKET]: <T id="transaction.type.ticket" m="Purchased" />,
  [txTypes.VOTED]: <T id="transaction.type.voted" m="Purchased" />,
  [txTypes.REVOKED]: <T id="transaction.type.revoked" m="Purchased" />,
  [txTypes.VOTE]: <T id="transaction.type.vote" m="Voted" />,
  [txTypes.REVOCATION]: <T id="transaction.type.revocation" m="Revoked" />,
  [txTypes.UNMINED]: <T id="transaction.type.unmined" m="Unmined" />,
  [txTypes.IMMATURE]: <T id="transaction.type.immature" m="Immature" />,
  [txTypes.MISSED]: <T id="transaction.type.missed" m="Missed" />,
  [txTypes.EXPIRED]: <T id="transaction.type.expired" m="Expired" />,
  [txTypes.LIVE]: <T id="transaction.type.live" m="Live" />
};
