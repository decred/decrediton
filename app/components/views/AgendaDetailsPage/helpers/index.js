export { default as Header } from "./Header";
export { default as VoteSection } from "./VoteSection";
export { default as AgendaCard } from "./AgendaCard";
