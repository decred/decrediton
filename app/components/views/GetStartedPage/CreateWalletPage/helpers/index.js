export { default as Container } from "./Container";
export { default as TitleWrapper } from "./TitleWrapper";
export { default as Section } from "./Section";
export { default as SeedArea } from "./SeedArea";
export { default as SeedWord } from "./SeedWord";
export { default as SeedError } from "./SeedError";
export { default as ButtonsBar } from "./ButtonsBar";
