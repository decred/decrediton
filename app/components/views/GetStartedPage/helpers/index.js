export { default as Label } from "./Label";
export { default as Input } from "./Input";
export { default as FormContainer } from "./FormContainer";
export { default as Row } from "./Row";
export { default as ButtonsBar } from "./ButtonsBar";
export { default as BackButtonArea } from "./BackButtonArea";
export { default as BackButton } from "./BackButton";
export { default as Content } from "./Content";
export { default as Title } from "./Title";
export { default as SubTitle } from "./SubTitle";
export { default as ContentContainer } from "./ContentContainer";
