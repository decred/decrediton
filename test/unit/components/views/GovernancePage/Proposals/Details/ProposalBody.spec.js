import { ProposalBody } from "views/ProposalDetailsPage/helpers";
import { render } from "@testing-library/react";
import { screen } from "@testing-library/react";
import fullMarkdown from "data/FullProposalMarkdown";

test("ProposalBody renders simple markdown", () => {
  const body = render(
    <ProposalBody body={"this is a *test* proposal description"} />
  );

  expect(body.container).toMatchSnapshot();
});

test("ProposalBody renders paragraphs", () => {
  render(<ProposalBody body={"First Para\n\nSecond Para"} />);
  expect(screen.getByText("First Para")).toBeInTheDocument();
  expect(screen.getByText("Second Para")).toBeInTheDocument();
});

test("ProposalBody does not render reference images", () => {
  const body = render(
    <ProposalBody body={"![Alt text][ref0]\n\n[ref0] url/to/image.jpg"} />
  );

  expect(body.container.querySelector("img")).toBeNull();
});

test("ProposalBody does not render inline images", () => {
  const body = render(<ProposalBody body={"![Alt text](/path/to/img.jpg)"} />);

  expect(body.container.querySelector("img")).toBeNull();

  const body2 = render(
    <ProposalBody body={'![Alt text](/path/to/img.jpg "Optional title")'} />
  );

  expect(body2.container.querySelector("img")).toBeNull();
});

test("ProposalBody renders full markdown", () => {
  const body = render(<ProposalBody body={fullMarkdown} />);

  expect(body.container).toMatchSnapshot();
});

test("ProposalBody skips html", () => {
  const body = render(<ProposalBody body={"this <b>html</b> is blocked"} />);

  expect(body.container).toMatchSnapshot();
});

test("ProposalBody protects links", () => {
  const body = render(<ProposalBody body={"[link](http://decred.org)"} />);

  expect(body.container.querySelector("a").getAttribute("href")).toEqual("#");
});
