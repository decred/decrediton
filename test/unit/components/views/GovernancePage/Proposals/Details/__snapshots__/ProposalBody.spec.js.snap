// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProposalBody renders full markdown 1`] = `
<div>
  <div
    class="markdown-body"
  >
    <p>
      Full markdown for testing
    </p>
    

    <hr />
    

    <h1>
      h1 Heading 8-)
    </h1>
    

    <h2>
      h2 Heading
    </h2>
    

    <h3>
      h3 Heading
    </h3>
    

    <h4>
      h4 Heading
    </h4>
    

    <h5>
      h5 Heading
    </h5>
    

    <h6>
      h6 Heading
    </h6>
    

    <h2>
      Horizontal Rules
    </h2>
    

    <hr />
    

    <hr />
    

    <hr />
    

    <h2>
      Typographic replacements
    </h2>
    

    <p>
      Enable typographer option to see result.
    </p>
    

    <p>
      (c) (C) (r) (R) (tm) (TM) (p) (P) +-
    </p>
    

    <p>
      test.. test... test..... test?..... test!....
    </p>
    

    <p>
      !!!!!! ???? ,,  -- ---
    </p>
    

    <p>
      "Smartypants, double quotes" and 'single quotes'
    </p>
    

    <h2>
      Emphasis
    </h2>
    

    <p>
      <strong>
        This is bold text
      </strong>
    </p>
    

    <p>
      <strong>
        This is bold text
      </strong>
    </p>
    

    <p>
      <em>
        This is italic text
      </em>
    </p>
    

    <p>
      <em>
        This is italic text
      </em>
    </p>
    

    <p>
      <del>
        Strikethrough
      </del>
    </p>
    

    <h2>
      Blockquotes
    </h2>
    

    <blockquote>
      

      <p>
        Blockquotes can also be nested...
      </p>
      

      <blockquote>
        

        <p>
          ...by using additional greater-than signs right next to each other...
        </p>
        

        <blockquote>
          

          <p>
            ...or with spaces between arrows.
          </p>
          

        </blockquote>
        

      </blockquote>
      

    </blockquote>
    

    <h2>
      Lists
    </h2>
    

    <p>
      Unordered
    </p>
    

    <ul>
      

      <li>
        Create a list by starting a line with 
        <code>
          +
        </code>
        , 
        <code>
          -
        </code>
        , or 
        <code>
          *
        </code>
      </li>
      

      <li>
        Sub-lists are made by indenting 2 spaces:
        

        <ul>
          

          <li>
            Marker character change forces new list start:
            

            <ul>
              

              <li>
                Ac tristique libero volutpat at
              </li>
              

            </ul>
            

            <ul>
              

              <li>
                Facilisis in pretium nisl aliquet
              </li>
              

            </ul>
            

            <ul>
              

              <li>
                Nulla volutpat aliquam velit
              </li>
              

            </ul>
            

          </li>
          

        </ul>
        

      </li>
      

      <li>
        Very easy!
      </li>
      

    </ul>
    

    <p>
      Ordered
    </p>
    

    <ol>
      

      <li>
        

        <p>
          Lorem ipsum dolor sit amet
        </p>
        

      </li>
      

      <li>
        

        <p>
          Consectetur adipiscing elit
        </p>
        

      </li>
      

      <li>
        

        <p>
          Integer molestie lorem at massa
        </p>
        

      </li>
      

      <li>
        

        <p>
          You can use sequential numbers...
        </p>
        

      </li>
      

      <li>
        

        <p>
          ...or keep all the numbers as 
          <code>
            1.
          </code>
        </p>
        

      </li>
      

    </ol>
    

    <p>
      Start numbering with offset:
    </p>
    

    <ol
      start="57"
    >
      

      <li>
        foo
      </li>
      

      <li>
        bar
      </li>
      

    </ol>
    

    <h2>
      Code
    </h2>
    

    <p>
      Inline 
      <code>
        code
      </code>
    </p>
    

    <p>
      Indented code
    </p>
    

    <pre>
      <code>
        // Some comments
line 1 of code
line 2 of code
line 3 of code

      </code>
    </pre>
    

    <p>
      Block code "fences"
    </p>
    

    <pre>
      <code>
        Sample text here...

      </code>
    </pre>
    

    <p>
      Syntax highlighting
    </p>
    

    <pre>
      <code
        class="language-js"
      >
        var foo = function (bar) {
  return bar++;
};

console.log(foo(5));

      </code>
    </pre>
    

    <h2>
      Tables
    </h2>
    

    <table>
      <thead>
        <tr>
          <th>
            Option
          </th>
          <th>
            Description
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            data
          </td>
          <td>
            path to data files to supply the data that will be passed into templates.
          </td>
        </tr>
        <tr>
          <td>
            engine
          </td>
          <td>
            engine to be used for processing templates. Handlebars is the default.
          </td>
        </tr>
        <tr>
          <td>
            ext
          </td>
          <td>
            extension to be used for dest files.
          </td>
        </tr>
      </tbody>
    </table>
    

    <p>
      Right aligned columns
    </p>
    

    <table>
      <thead>
        <tr>
          <th
            style="text-align: right;"
          >
            Option
          </th>
          <th
            style="text-align: right;"
          >
            Description
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            style="text-align: right;"
          >
            data
          </td>
          <td
            style="text-align: right;"
          >
            path to data files to supply the data that will be passed into templates.
          </td>
        </tr>
        <tr>
          <td
            style="text-align: right;"
          >
            engine
          </td>
          <td
            style="text-align: right;"
          >
            engine to be used for processing templates. Handlebars is the default.
          </td>
        </tr>
        <tr>
          <td
            style="text-align: right;"
          >
            ext
          </td>
          <td
            style="text-align: right;"
          >
            extension to be used for dest files.
          </td>
        </tr>
      </tbody>
    </table>
    

    <h2>
      Links
    </h2>
    

    <p>
      <span>
        <a
          href="#"
        >
          link text
        </a>
      </span>
    </p>
    

    <p>
      <span>
        <a
          href="#"
        >
          link with title
        </a>
      </span>
    </p>
    

    <p>
      Autoconverted link 
      <span>
        <a
          href="#"
        >
          https://github.com/nodeca/pica
        </a>
      </span>
       (enable linkify to see)
    </p>
    

    <h2>
      Images
    </h2>
    

    <p>
      <span>
        Minion
      </span>
      

      <span>
        Stormtroopocat
      </span>
    </p>
    

    <p>
      Like links, Images also have a footnote style syntax
    </p>
    

    <p>
      <span>
        Alt text
      </span>
    </p>
    

    <p>
      With a reference later in the document defining the URL location:
    </p>
    

    <h2>
      Plugins
    </h2>
    

    <p>
      The killer feature of 
      <code>
        markdown-it
      </code>
       is very effective support of

      <span>
        <a
          href="#"
        >
          syntax plugins
        </a>
      </span>
      .
    </p>
    

    <h3>
      <span>
        <a
          href="#"
        >
          Emojies
        </a>
      </span>
    </h3>
    

    <blockquote>
      

      <p>
        Classic markup: :wink: :crush: :cry: :tear: :laughing: :yum:
      </p>
      

      <p>
        Shortcuts (emoticons): :-) :-( 8-) ;)
      </p>
      

    </blockquote>
    

    <p>
      see 
      <span>
        <a
          href="#"
        >
          how to change output
        </a>
      </span>
       with twemoji.
    </p>
    

    <h3>
      <span>
        <a
          href="#"
        >
          Subscript
        </a>
      </span>
       / 
      <span>
        <a
          href="#"
        >
          Superscript
        </a>
      </span>
    </h3>
    

    <ul>
      

      <li>
        19^th^
      </li>
      

      <li>
        H
        <del>
          2
        </del>
        O
      </li>
      

    </ul>
    

    <h3>
      <span>
        <a
          href="#"
        >
          &lt;ins&gt;
        </a>
      </span>
    </h3>
    

    <p>
      ++Inserted text++
    </p>
    

    <h3>
      <span>
        <a
          href="#"
        >
          &lt;mark&gt;
        </a>
      </span>
    </h3>
    

    <p>
      ==Marked text==
    </p>
    

    <h3>
      <span>
        <a
          href="#"
        >
          Footnotes
        </a>
      </span>
    </h3>
    

    <p>
      Footnote 1 link[^first].
    </p>
    

    <p>
      Footnote 2 link[^second].
    </p>
    

    <p>
      Inline footnote^[Text of inline footnote] definition.
    </p>
    

    <p>
      Duplicated footnote reference[^second].
    </p>
    

    <p>
      [^first]: Footnote 
      <strong>
        can have markup
      </strong>
    </p>
    

    <pre>
      <code>
        and multiple paragraphs.

      </code>
    </pre>
    

    <p>
      [^second]: Footnote text.
    </p>
    

    <h3>
      <span>
        <a
          href="#"
        >
          Definition lists
        </a>
      </span>
    </h3>
    

    <p>
      Term 1
    </p>
    

    <p>
      :   Definition 1
with lazy continuation.
    </p>
    

    <p>
      Term 2 with 
      <em>
        inline markup
      </em>
    </p>
    

    <p>
      :   Definition 2
    </p>
    

    <pre>
      <code>
            { some code, part of Definition 2 }

Third paragraph of definition 2.

      </code>
    </pre>
    

    <p>
      <em>
        Compact style:
      </em>
    </p>
    

    <p>
      Term 1
~ Definition 1
    </p>
    

    <p>
      Term 2
~ Definition 2a
~ Definition 2b
    </p>
    

    <h3>
      <span>
        <a
          href="#"
        >
          Abbreviations
        </a>
      </span>
    </h3>
    

    <p>
      This is HTML abbreviation example.
    </p>
    

    <p>
      It converts "HTML", but keep intact partial entries like "xxxHTMLyyy" and so on.
    </p>
    

    <p>
      *[HTML]: Hyper Text Markup Language
    </p>
    

    <h3>
      <span>
        <a
          href="#"
        >
          Custom containers
        </a>
      </span>
    </h3>
    

    <p>
      ::: warning

      <em>
        here be dragons
      </em>
      
:::
    </p>
    

    <hr />
    

    <h1>
      End of standard markdown test
    </h1>
    

    <h1>
      Start of custom markdown test
    </h1>
    

    <p>
      <span>
        politeia gui image
      </span>
    </p>
    

    <p>
      <span>
        politeia gui ref image
      </span>
    </p>
    

    <p>
      and the reference is right below this
    </p>
    

    <p>
       Schnorr Signatures
    </p>
    

    

    

    <p>
      exp/*
    </p>
    

    <p>
      ¼script¾alert(¢XSS¢)¼/script¾
    </p>
    

    <hr />
    

    <p>
      Let's add some unicode emojis, just for fun!
    </p>
    

    <p>
      ☯ 262F yin yang
☢ 2622 radioactive sign (trefoil)
⚽ 26BD soccer
    </p>
    

    <h1>
      ⚖️
    </h1>
  </div>
</div>
`;

exports[`ProposalBody renders simple markdown 1`] = `
<div>
  <div
    class="markdown-body"
  >
    <p>
      this is a 
      <em>
        test
      </em>
       proposal description
    </p>
  </div>
</div>
`;

exports[`ProposalBody skips html 1`] = `
<div>
  <div
    class="markdown-body"
  >
    <p>
      this 
      html
       is blocked
    </p>
  </div>
</div>
`;
